org-ql.el

```diff
@@ -454,38 +459,38 @@ Returns cons (INHERITED-TAGS . LOCAL-TAGS)."
                  org-ql-tags-cache))
       (puthash position all-tags tags-cache))))
 
-(defun org-ql--expand-tag-hierarchy (tags &optional excluded)
+(defun org-ql--expand-tag-hierarchy (tags &optional groups)
   "Return TAGS along with their associated group tags.
-This function recursively searches for groups that each given tag belongs to,
-directly or indirectly, and includes the corresponding group tags to the result.
+This function recursively searches for groups in GROUPS that
+each tag belongs to and includes the corresponding group tags to the result.
 
 TAGS should be a list of tags (i.e., strings).
-If non-nil, EXCLUDED should be a list of group tags that will not be
-automatically added to the results unless they are already in TAGS."
-  (let ((groups (org-tag-alist-to-groups org-current-tag-alist))
-        (excluded (append tags excluded))
-        group-tags)
+If GROUPS is non-nil, then it must be a list of tag group definitions of
+the form (tag member1 member2 ...).  Otherwise, it defaults to the buffer's
+current tag hierarchy."
+  (let ((groups (or groups (org-tag-alist-to-groups org-current-tag-alist)))
+        matching-group-tags)
+    ;; Remove groups with tags already in the results.
+    (mapc (lambda (tag) (setq groups (assoc-delete-all tag groups))) tags)
```

Comment:
I don't fully understand the values these data structures have at runtime, because I haven't stepped through the code.  But it seems to me like this step is something that ought to be avoided.  If it seems necessary here, it might be because the design needs to be adjusted.  It seems to me like this is work that shouldn't need to be done every time this function is called.

Reply:
In the above snippet, `groups` essentially holds the current set of group definitions (in the first call, it should be the result of `(org-tag-alist-to-groups org-current-tag-alist)`. In order to avoid iterating over the members of groups whose group-tag is already in the tags passed to this function, I simply delete them from this variable. I think that this is an improvement over the previous implementation because if this function ends up recursing, the number of iterations of the `pcase-dolist` loop will be definitely less.